src/net.cpp

```diff
@@ -1152,6 +1152,37 @@ void CConnman::AcceptConnection(const ListenSocket& hListenSocket) {
     RandAddEvent((uint32_t)id);
 }
 
+bool CConnman::AddConnection(const std::string& address, const ConnectionType conn_type)
+{
+    assert(conn_type == ConnectionType::OUTBOUND_FULL_RELAY || conn_type == ConnectionType::BLOCK_RELAY);
+
+    int outbound_full_relay = 0;
+    int outbound_block_relay = 0;
+
+    {
+        // count existing connections for each type
+        LOCK(cs_vNodes);
+        for (const CNode* pnode : vNodes) {
+            if (pnode->m_conn_type == ConnectionType::OUTBOUND_FULL_RELAY) {
+                ++outbound_full_relay;
+            } else if (pnode->m_conn_type == ConnectionType::BLOCK_RELAY) {
+                ++outbound_block_relay;
+            }
+        }
+    }
+
+    // max connections of specified type already exist
+    if (conn_type == ConnectionType::OUTBOUND_FULL_RELAY && outbound_full_relay >= m_max_outbound_full_relay) return false;
+    if (conn_type == ConnectionType::BLOCK_RELAY && outbound_block_relay >= m_max_outbound_block_relay) return false;
+
+    // max total outbound connections already exist
+    CSemaphoreGrant grant(*semOutbound, true);
+    if (!grant) return false;
+
+    OpenNetworkConnection(CAddress(), false, &grant, address.c_str(), conn_type);
```

Comment:
There are several reasons that `OpenNetworkConnections()` can fail (eg a connection to that peer already exists, we fail to open a socket, etc). It'd be nice to change `OpenNetworkConnection()` to return a boolean and return that to the RPC caller.

If we don't return this value, then the RPC looks like it succeeded even if a connection has not been opened:

```
→ .bitcoin-cli addconnection i-am-not-an-address outbound
{
  "address": "i-am-not-an-address",
  "connectiontype": "outbound"
}
→ ./bitcoin-cli getpeerinfo
[
]
```

Reply:
`./bitcoin-cli addnode i-am-not-an-address onetry` gives the same result, as does trying with a valid but unreachable address. Even with NET logging enabled, I only see an error in debug.log for the unreachable address ("connection to XXX timeout"). Having a more reliable indication *somewhere* of what went wrong would probably be a good idea both here and for addnode might be good. Not sure that needs to be in this PR though, behaving no worse than the existing addnode doesn't seem too bad.

Reply:
Agree that this can be done separately if the behaviour here is the same as for `addnode`.

Reply:
agree with the idea. I've started a branch to incorporate: https://github.com/amitiuttarwar/bitcoin/commits/2020-11-open-conn-improvements. I want to add some testing and review the failures / think through logging. but will do this as a follow up.

Reply:
0c4d673716f2eb37efdb81ad2d7d2

```suggestion
    OpenNetworkConnection(CAddress(), /* fCountFailure */ false, &grant, address.c_str(), conn_type);
```

Reply:
I agree with returning a boolean; it may also be convenient for it to return the peer id of the connection via an out param.

Reply:
resolving this conversation, tracking better return value from OpenNetworkConnection as a future improvement in https://github.com/bitcoin/bitcoin/pull/19315#issuecomment-735282520

Reply:
Filed an issue for this at #20552